stl/inc/xstring

```diff
@@ -191,23 +272,45 @@ struct _WChar_traits { // char_traits for the char16_t-likes: char16_t, wchar_t,
         if constexpr (is_same_v<_Elem, wchar_t>) {
             return __builtin_wmemchr(_First, _Ch, _Count);
         } else {
-            return _Char_traits<_Elem, unsigned short>::find(_First, _Count, _Ch);
+            return _Primary_char_traits::find(_First, _Count, _Ch);
         }
 #else // _HAS_CXX17
         return reinterpret_cast<const _Elem*>(_CSTD wmemchr(reinterpret_cast<const wchar_t*>(_First), _Ch, _Count));
 #endif // _HAS_CXX17
     }
 
-    static _Elem* move(_Out_writes_all_(_Count) _Elem* const _First1, _In_reads_(_Count) const _Elem* const _First2,
-        const size_t _Count) noexcept /* strengthened */ {
+    static _CONSTEXPR20 _Elem* move(_Out_writes_all_(_Count) _Elem* const _First1,
+        _In_reads_(_Count) const _Elem* const _First2, const size_t _Count) noexcept /* strengthened */ {
         // copy [_First2, _First2 + _Count) to [_First1, ...)
-        return reinterpret_cast<_Elem*>(
-            _CSTD wmemmove(reinterpret_cast<wchar_t*>(_First1), reinterpret_cast<const wchar_t*>(_First2), _Count));
+#if _HAS_MEMCPY_MEMMOVE_INTRINSICS
+        __builtin_memmove(_First1, _First2, _Count * sizeof(_Elem));
+#else // ^^^ _HAS_MEMCPY_MEMMOVE_INTRINSICS ^^^ / vvv !_HAS_MEMCPY_MEMMOVE_INTRINSICS vvv
+#ifdef __cpp_lib_is_constant_evaluated
+        if (_STD is_constant_evaluated()) {
+            return _Primary_char_traits::move(_First1, _First2, _Count);
+        }
+#endif // __cpp_lib_is_constant_evaluated
+
+        _CSTD wmemmove(reinterpret_cast<wchar_t*>(_First1), reinterpret_cast<const wchar_t*>(_First2), _Count);
+#endif // ^^^ !_HAS_MEMCPY_MEMMOVE_INTRINSICS ^^^
+
+        return _First1;
     }
 
-    static _Elem* assign(_Out_writes_all_(_Count) _Elem* const _First, size_t _Count, _Elem _Ch) noexcept
+    static _CONSTEXPR20 _Elem* assign(
+        _Out_writes_all_(_Count) _Elem* const _First, size_t _Count, const _Elem _Ch) noexcept
     /* strengthened */ {
         // assign _Count * _Ch to [_First, ...)
+#ifdef __cpp_lib_is_constant_evaluated
+        if (_STD is_constant_evaluated()) {
+            for (_Elem* _Next = _First; _Count > 0; --_Count, ++_Next) {
+                *_Next = _Ch;
+            }
+
+            return _First;
+        }

```

Comment:
This looks like the same code contained in `_Char_traits::assign` function.  Can we substitute with `_Primary_char_traits::assign(_First, _Count, _Ch)`, similar to what is done in the `copy` and `move` functions?

Reply:
Yep - during constant evaluation, there's no reason to avoid calling helper functions. (At runtime, there's the minor concern of debug codegen.) I'll make this change, thanks!